Shared element transitions: Changes for new API

jakearchibald opened the pull request:
Changes from https://github.com/WICG/shared-element-transitions/issues/175#issuecomment-1210872202

Note: This is based on https://github.com/tabatkins/specs/pull/96

jakearchibald committed:
Changes for new API

vmpstr commented on css-shared-element-transitions/index.bs:
```diff
@@ -324,103 +320,124 @@ A {{Document}} additionally has:
 
-callback PrepareCallback = Promise<any> ();
+dictionary DOMTransitionInit {
+    ChangeDOMCallback changeDOM;
```
```suggestion
    ChangeDOMCallback updateDOM;
```
to align with "domUpdated" promise below

khushalsagar replied:
+1.

vmpstr commented on css-shared-element-transitions/index.bs:
```diff
@@ -324,103 +320,124 @@ A {{Document}} additionally has:
-    : <dfn>phase</dfn>
-    :: "`idle`", "`preparing`", "`running`", or "`finished`".
-        Initially "`idle`".
+    : <dfn>concluded</dfn>
```
I think this is fine, but why not something like "done" or "finished"? (https://www.w3.org/TR/design-principles/#naming-common-words)

jakearchibald replied:
I thought "done" and "finished" imply 'successfully' whereas "concluded" didn't. I could use "settled" which is the language promises use.

vmpstr replied:
Makes sense, I don't really have much of a preference just wanted to see if there was a rationale for using 'concluded'

khushalsagar replied:
Another suggestion: "active". Initially true and flips to false.

Otherwise between concluded and settled, I like concluded.

vmpstr commented on css-shared-element-transitions/index.bs:
```diff
@@ -547,35 +566,26 @@ The {{SameDocumentTransition/finished}} [=getter steps=] are to return [=this's=
                 although the render steps in the HTML spec act as if it's synchronous.
 
-        1. Let |timedOut| be false.
+        1. If |transition| has [=DOMTransition/concluded=], then abort these steps.
```
which steps? the following steps?

jakearchibald replied:
"abort these steps" is pretty common spec language to mean exit from the current algorithm steps. It behaves similar to `return`, but for cases where no return value is possible.

vmpstr commented on css-shared-element-transitions/index.bs:
```diff
@@ -692,37 +698,43 @@ The {{SameDocumentTransition/finished}} [=getter steps=] are to return [=this's=
+
+    1. [=Assert=]: |transition| has not [=DOMTransition/concluded=].
+
+    1. If |transition|'s [=DOMTransition/DOM change callback called=] is false, then [=call the DOM change callback=] of |transition|.
```
I think we need to clarify timings of things. These callbacks are asynchronous, and I'm not sure we should just call the callback and forget about it anytime a transition is aborted, since that has a potential of calling several callbacks before any one of the is finished, right? 

I was under the impression that we should queue the callback into some queue that is always calling the next callback, _waits for the returned promise to settle_, and then call the next callback. That may have a potential of some sort of a deadlock i'm sure, but that would be a developer error.

WDYT?

jakearchibald replied:
I brought this up in https://github.com/WICG/shared-element-transitions/issues/175#issuecomment-1220817698

> While writing (the PR), I think we got something wrong in our plan:
> 
> > We need the pending callbacks to run before capturing outgoing state for the new transition.
> 
> I think this violates our "keep the transition and navigation separate" rule. Something other than our API should be in charge of queuing callbacks that should be queued.
> 
> The problem comes if a callback takes a long time to resolve. We would either:
> 
> * Queue indefinitely, meaning one bad callback can prevent all future queued DOM changes on the page.
> * Timeout. But, in that case, the callback could still make DOM changes, so we're still allowing these callbacks to run in parallel.
> 
> Neither of these are good.
> 
> The navigation API doesn't perform any queuing, it relies on developers to manage that themselves. It feels like we should follow the same model.

jakearchibald replied:
I guess the key question is: is it our responsibility to queue these things, or is that something an asynchronous DOM system will already be doing, and we'd just get in the way by doing it again.

vmpstr replied:
Yeah, I agree that this is the question. My impression from our discussion was that the API would be responsible for serializing the callbacks. It seems like that may be problematic and does not align with navigation API. We likely should discuss this briefly to understand that timing of these callbacks, specifically their completion, would then depend on the callback duration itself, meaning it's fairly arbitrary

jakearchibald replied:
Yes, in our discussion we were definitely set on queuing. I'm just not sure we thought the deadlock thing through. Or, at least, I hadn't.

khushalsagar replied:
Could you give an example of a case where this could result in deadlocks? If the callbacks are being executed in the order createTransition is called, why would the first callback need to depend on any operation in the second callback.

@vmpstr and I discussed 3 options for this:

1. Execute the callbacks concurrently and while all steps for the transition are being executed. Since these callbacks are updating the DOM we could end up snapshotting a random frame and/or have the callbacks update the DOM midway through a transition. That seems pretty bad.
2. Execute the callbacks concurrently but wait for promises to settle before snapshotting. This avoids the latency of everything executing serially but the execution within each of them will interleave. But it ensures we're in a stable state before initiating capture and rest of the algorithm steps.
3. Execute the callbacks serially, waiting for each's promise to settle. And then initiate the transition. Sounds like this has the possibility for deadlocks which I'd like to understand better.

The first one seems pretty bad since we'll be running callbacks that are intended to update the DOM during the transition. Both 2 or 3 would be fine. If the developer doesn't like the default they can always build a queue in user land to switch to the other behaviour.

jakearchibald replied:
> Could you give an example of a case where this could result in deadlocks?

```js
document.createTransition({
  updateDOM: () => new Promise(() => {}),
})
```

This might be done in code from LibraryA, but it means that all transitions are deadlocked from now on.

> 1. Execute the callbacks concurrently and while all steps for the transition are being executed. Since these callbacks are updating the DOM we could end up snapshotting a random frame and/or have the callbacks update the DOM midway through a transition. That seems pretty bad.

Here's where my head's at:

**Our API _enhances_ DOM updates with transitions.** Transitions aren't an essential part of the process, but the DOM update is. This is the thinking that led us to "the callback should always run, even if the transition cannot".

**Most DOM updates are synchronous, or queued behind a microtask.** In that case, we're the ones that make it async, due to the timing of the `updateDOM` callback. But even then, in these cases the DOM updates will complete in sequence.

**If the DOM updates are async regardless of transitions, the concurrency problem needs to be solved regardless of transitions.** If your async DOM framework breaks due to its own asyncness, then it isn't our responsibility to fix that. In fact, if we try and fix it, we're probably just getting in the way of the DOM frameworks that are correctly managing their own asyncness.

The Navigation API for instance doesn't provide a scheduler. It will allow many [intercept handlers](https://wicg.github.io/navigation-api/#ref-for-dom-navigateevent-intercept%E2%91%A0%E2%91%A3) to run concurrently. It provides an [abort signal](https://wicg.github.io/navigation-api/#ref-for-dom-navigateevent-signal%E2%91%A0), and trusts that the developer does the right thing to avoid overlapping actions.

I don't think we should provide a similar system, because skipping the transition doesn't always mean skipping the DOM change. It does in some cases, but that's judged by some other system (like the Navigation API).

So, if we call a second callback before the promise from the first has settled, it's only a problem if it would also be a problem without our API. So, it's not our problem to solve.

I might be missing part of the picture though!

khushalsagar replied:
> This might be done in code from LibraryA, but it means that all transitions are deadlocked from now on.

Ah, so this would be the case of a developer error. We should limit the wait for the promise on a timeout, similar to what we do when dispatching this callback for the transition.

> If the DOM updates are async regardless of transitions, the concurrency problem needs to be solved regardless of transitions.

Let's say the developer is using the same callback outside of transitions. An event occurs which initiates the transition to state A and developer code issues the callback. This callback yields because of an async event (like disk fetch) and then another event happens which wants to transition to state B. The developer will already need to build a system to either queue the callback or handle the affect that they can be concurrent. That part SGTM.

But let's assume that the concurrency is handled by the developer code so its okay to run them in parallel. I think the developer would still want the DOM to reach a stable state, in this case to state A, before the browser snapshots the before state. If we go with the behaviour in 1) then they must build their own system to ensure all pending callbacks have finished executing before calling createTransition(...) again. With 2) as the default behaviour, if the developer doesn't want the pending callbacks to block then they can track the pending transition and use the skipTransition() API.

jakearchibald replied:
> We should limit the wait for the promise on a timeout, similar to what we do when dispatching this callback for the transition.

But we can't be certain that it's 'stuck', it might just be slow. So, if we timeout, we could potentially be allowing two callbacks to run concurrently. And, if we're fine with that, why not just _always_ run them concurrently. That's more reliable for developers.

This is the case that made me change my mind about queuing.

> But let's assume that the concurrency is handled by the developer code so its okay to run them in parallel. I think the developer would still want the DOM to reach a stable state, in this case to state A, before the browser snapshots the before state. If we go with the behaviour in 1) then they must build their own system to ensure all pending callbacks have finished executing before calling createTransition(...) again. With 2) as the default behaviour, if the developer doesn't want the pending callbacks to block then they can track the pending transition and use the skipTransition() API.

So the code would be something like this:

```js
async function runUpdate() {
  const data = await fetchData();

  const transition = document.createTransition({
    async updateDOM() {
      await updateDOMSomehow(data);
      await loadFontsWithAgressiveTimeout();
    },
  });

  await transition.domUpdated;
}
```

Let's assume that the underlying framework can handle two concurrent DOM updates via `updateDOMSomehow`. I think it's fair to say it would only allow these to run concurrently if they're completely separate parts of the page. Otherwise, I'd expect the framework to either queue the `updateDOMSomehow` actions, or reject the first in favour of the last. So let's say the transitions are happening in ComponentA and ComponentB respectively.

The ideal here is probably two non-root transitions that can also run at the same time. But ignoring that…

The developer intends to perform two transitions. Snapping visual state and then transitioning is kinda ugly, so we're already in non-ideal territory. Given the lack of non-root transitions, the smoothest experience is probably to queue the transitions so each can run, but that isn't what the developer is doing in the code above (but our API does make it possible).

So any of the following could happen:

- Switch to ComponentA new state, transition to ComponentB state.
- Skip ComponentA new state transition mid-way, transition to ComponentB state.
- Transition for ComponentA is skipped, but it hasn't finished changing the DOM yet. Transition to ComponentB state. Mid-way through, ComponentA switches to new state.

Queuing the `updateDOM` callbacks prevents that third possibility. But, are we really preventing something undesirable?

jakearchibald replied:
TODO:

- Add a note that changes are expected to be syncish, or the framework deals with it
- Don't need to get all transitions, because there'll be 0 or 1. Just store that on the doc.

khushalsagar commented on css-shared-element-transitions/index.bs:
```diff
@@ -324,103 +320,124 @@ A {{Document}} additionally has:
 
 <dl dfn-for="document">
     : <dfn>pending same document transition outgoing capture</dfn>
```
```suggestion
    : <dfn>pending DOM transition outgoing capture</dfn>
```

for consistency with the class name.

khushalsagar commented on css-shared-element-transitions/index.bs:
```diff
@@ -324,103 +320,124 @@ A {{Document}} additionally has:
-    :: a {{SameDocumentTransition}} or null. Initially null.
+    :: a {{DOMTransition}} or null. Initially null.
 
     : <dfn>running same document transition</dfn>
```
```suggestion
    : <dfn>running DOM transition</dfn>
```

khushalsagar commented on css-shared-element-transitions/index.bs:
```diff
@@ -324,103 +320,124 @@ A {{Document}} additionally has:
+    :: a {{DOMTransition}} or null. Initially null.
+
+    : <dfn>transition suppressing rendering</dfn>
+    :: a {{DOMTransition}} or null. Initially null.
```
IIUC only one of these can be set at a time. Maybe leave a note indicating that?

In code we just have one member on Document and its "phase" will indicate what stage its at and which algorithm needs to be executed. The spec could also do that but I'm fine either way.

jakearchibald replied:
Makes more sense to have one thing.

khushalsagar commented on css-shared-element-transitions/index.bs:
```diff
@@ -324,103 +320,124 @@ A {{Document}} additionally has:
+    ChangeDOMCallback changeDOM;
+};
+
+callback ChangeDOMCallback = Promise<any> ();
```
```suggestion
callback UpdateDOMCallback = Promise<any> ();
```

If you end up going with the suggestion above.

khushalsagar commented on css-shared-element-transitions/index.bs:
```diff
@@ -324,103 +320,124 @@ A {{Document}} additionally has:
+
+    1. [=list/For each=] |transitionToSkip| of |transitionsToSkip|, [=skip the page transition=] |transitionToSkip| with an "{{AbortError}}" {{DOMException}} in [=this's=] [=relevant Realm=].
+
+    1. [=Assert=]: |document|'s [=document/pending same document transition outgoing capture=] is null.
```
Assert that all 3 DOMTransitions on the document are null?

khushalsagar commented on css-shared-element-transitions/index.bs:
```diff
@@ -643,7 +649,7 @@ The {{SameDocumentTransition/finished}} [=getter steps=] are to return [=this's=
 
     <pre highlight=js>
-    new SameDocumentTransition()
+    new DOMTransition()
```
Doesn't this need to be updated to use the new document.createTransition(...) syntax? Or is the idea to keep both. I assumed we were only doing the createTransition one to begin with.

jakearchibald replied:
Oof, yeah, I just forgot about this

khushalsagar commented on css-shared-element-transitions/index.bs:
```diff
@@ -324,103 +320,124 @@ A {{Document}} additionally has:
-    1. If [=this's=] [=SameDocumentTransition/phase=] is not "`finished`",
-        then [=abandon the page transition=] [=this=]
+    1. If [=this=] has not [=DOMTransition/concluded=],
+        then [=skip the page transition=] for [=this=]
```
Maybe we add a param to allow skipping the transition without running the callback if that ends up as a common use-case.

jakearchibald replied:
If we stick with the naming here, I think it'd need to be a different method. Like `abandon` 😄

khushalsagar commented on css-shared-element-transitions/index.bs:
```diff
@@ -978,3 +996,20 @@ The {{SameDocumentTransition/finished}} [=getter steps=] are to return [=this's=
+
+    1. [=Assert=]: |transition|'s [=DOMTransition/DOM change callback called=] is false.
+
+    1. Let |callbackPromise| be the result of [=/invoking=] |transition|'s [=DOMTransition/DOM change callback=].
```
Why not set DOMTransition/DOM updated promise to |callbackPromise|?

jakearchibald replied:
That would break this case:

```js
const transition = document.createTransition({
  updateDOM() {
    // …
  }
});

await transition.domUpdated;
console.log('updated!');
```

`'updated!'` would never log, because `transition.domUpdated` never settles. Some other promise that `transition.domUpdated` is later set to is the one that settles.

khushalsagar replied:
Oh I get it now. Coz we haven't even run the callback yet so we'll end up switching the promise when that happens. SGTM then.

jakearchibald committed:
Changes following review

jakearchibald commented:
Summary of changes:

- Added a "Transitions as an enhancement" section, which details how this API is about transitions only, and that DOM changes always take priority.
- Phases on {{DOMTransition}} return!
- Added some prose so we can refer to the phases as a sequence.
- The document now just has one "active DOM transition". The phase of this is used to call the capture or update steps on each render.
- The document's "transition suppressing rendering" is now a boolean, and I've added a stub monkey-patch to define what that actually does.
- `changeDOM` to `updateDOM`
- `ChangeDOMCallback` to `UpdateDOMCallback`
- 'concluded' is now 'done'. I think it's clear enough now it's a phase.
- Moved the example to the definition of `createTransition`
- Updated the examples
- `updateDOM` is required (pretty sure that was our intent anyway)

jakearchibald committed:
updateDOM is required

khushalsagar commented on css-shared-element-transitions/index.bs:
```diff
@@ -623,106 +703,47 @@ The {{SameDocumentTransition/finished}} [=getter steps=] are to return [=this's=
 
-    1. If |document|'s [=document/running same document transition=] is |transition|,
-        then set |document|'s [=document/running same document transition=] to null.
+    1. [=Reject=] |transition|'s [=DOMTransition/ready promise=] with |reason|.
```
Would we still do this if this promise was already resolved? If we skip during the animation phase.

jakearchibald replied:
Rejecting (and resolving) is a no-op if the promise has already rejected/resolved, but yeah, it'll probably be clearer to be explicit about that, or at least add a note.

khushalsagar commented on css-shared-element-transitions/index.bs:
```diff
@@ -438,34 +519,52 @@ The {{SameDocumentTransition/finished}} [=getter steps=] are to return [=this's=
+    1. If |document|'s [=document/active DOM transition=] is not null, then:
 
-        1. Set |document|'s [=document/pending same document transition outgoing capture=] to null.
+        1. If |document|'s [=document/active DOM transition=]'s [=DOMTransition/phase=] is "`pending-capture`", then [=perform an outgoing capture=] with |document|'s [=document/active DOM transition=].
```
I initially thought we need another phase between pending-capture and dom-update-callback-called. Since other tasks can run between the pending-capture finishing asynchronously and we don't want to run [=perform an outgoing capture=]twice.

But since document/transition suppressing rendering is set in perform an outgoing capture, update the rendering can only run once! We should add a note for that, it's quite subtle. :)

jakearchibald replied:
Ehhhh, it's way more subtle than I intended. I think we should add another phase for this to make it clearer.

jakearchibald replied:
(bless you for assuming I was being smart here rather than dumb 😄)

khushalsagar reviewed (approved):
The remaining couple of comments are minor. Please free to land.
